scylla/src/transport/connection.rs

```diff
@@ -772,11 +772,59 @@ impl Connection {
 
     pub(crate) async fn batch_with_consistency(
         &self,
-        batch: &Batch,
+        init_batch: &Batch,
         values: impl BatchValues,
         consistency: Consistency,
         serial_consistency: Option<SerialConsistency>,
     ) -> Result<QueryResult, QueryError> {
+        let batch = {
+            let mut to_prepare = Vec::<usize>::new();
+
+            {
+                let mut value_iter = values.batch_values_iter();
+                for (id, stmt) in init_batch.statements.iter().enumerate() {
+                    if let BatchStatement::Query(_) = stmt {
+                        let value = value_iter.next_serialized().transpose()?;
+                        if let Some(v) = value {
+                            if v.len() > 0 {
+                                to_prepare.push(id);
+                            }
+                        }
+                    } else {
+                        value_iter.skip_next();
+                    }
+                }
+            }
+
+            let mut batch: Batch = Default::default();
+
+            if to_prepare.is_empty() {
+                batch = init_batch.clone();
+            } else {
+                batch.config = init_batch.config.clone();
+
+                let mut vec_id = 0;
+                for (id, stmt) in init_batch.statements.iter().enumerate() {
+                    match stmt {
+                        BatchStatement::Query(query) => {
+                            if vec_id < to_prepare.len() && id == to_prepare[vec_id] {
+                                vec_id += 1;
+                                let prepared = self.prepare(query).await?;
+                                batch.append_statement(prepared.clone());
+                            } else {
+                                batch.append_statement(query.clone());
+                            }
+                        }
+                        BatchStatement::PreparedStatement(prepared) => {
+                            batch.append_statement(prepared.clone());
+                        }
+                    }
+                }
+            }
+
+            batch
+        };
```

Comment:
Finally, let's see if we can extract all of the added logic here to a separate method/function. It's getting quite long.

Reply:
I don't think you need to copy or clone `values`, you only need to pass it by reference. As I can see, it's only needed for `.batch_values_iter()` which takes `values` by reference - so you can pass `values` by reference to that function.

I imagine that the final method will look like this (I hope I didn't mess up lifetimes):

```rust
async fn prepare_batch<'b>(
    &self,
    batch: &'b Batch,
    values_iter: BatchValuesIterator<'_>,
) -> Cow<'b, Batch> {
    // ...
}
```

Reply:
I have a few errors and I have to change it a little bit to:
```rust
 async fn prepare_batch<'b>(
        &self,
        init_batch: &'b Batch,
        mut values_iter: impl BatchValuesIterator<'_>,
    ) -> Result<Cow<'b, Batch>, QueryError> {
    //...
}
```
but it still does not build, there is a problem with lifetimes in `session_test.rs:408`  (implementation of `std::marker::Send` is not general enough
    |
    = note: `std::marker::Send` would have to be implemented for the type `&'0 str`, for any lifetime `'0`...
    = note: ...but `std::marker::Send` is actually implemented for the type `&'1 str`, for some specific lifetime `'1`).
    I am not sure how to fix it

Reply:
I see. Please try passing `&values` instead. For some reason, the `BatchValuesIterator` cannot be held across await points, so I suppose there might be issues when trying to pass it to async functions.